pymepix/processing/logic/centroid_calculator.py

```diff
@@ -0,0 +1,235 @@
+
+        This is useful in reducing the computational time in centroiding and can filter out
+        noise. """
+        return self._tot_threshold.value
+
+    @tot_threshold.setter
+    def tot_threshold(self, tot_threshold):
+        self._tot_threshold.value = tot_threshold
+
+    @property
+    def triggers_processed(self):
+        """ Setting for the number of packets skiped during processing. Every packet_skip packet is processed. 
+        This means for a value of 1 every packet is processed. For 2 only every 2nd packet is processed. """
+        return self._triggers_processed.value
+
+    @triggers_processed.setter
+    def triggers_processed(self, triggers_processed):
+        self._triggers_processed.value = triggers_processed
+
+    def process(self, data):
+        if data is not None:
+            shot, x, y, tof, tot = self.__skip_triggers(*data)
+            chunks = self.__divide_into_chunks(shot, x, y, tof, tot)
+            centroids_in_chunks = self.perform_centroiding(chunks)
+
+            return self.__centroid_chunks_to_centroids(centroids_in_chunks)
+        else:
+            return None
+
+    def __skip_triggers(self, shot, x, y, tof, tot):
+        unique_shots = np.unique(shot)
+        selected_shots = unique_shots[::self.triggers_processed]
+        mask = np.isin(shot, selected_shots)
+        return shot[mask], x[mask], y[mask], tof[mask], tot[mask]
+
+    def __divide_into_chunks(self, shot, x, y, tof, tot):
+        # Reordering the voxels can have an impact on the clusterings result. See CentroidCalculator.perform_clustering string doc for further information!
+        order = shot.argsort()
+        shot, x, y, tof, tot = shot[order], x[order], y[order], tof[order], tot[order]
+        split_indices = self.__calc_trig_chunks_split_indices(shot)
+        if len(split_indices) > 0:
+
+            shot, x, y, tof, tot = [np.split(arr, split_indices) for arr in [shot, x, y, tof, tot]]
+
+            chunks = []
+            for i in range(len(shot)):
+                chunks.append((shot[i], x[i], y[i], tof[i], tot[i]))
+            return chunks
+        else:
+            return [(shot, x, y , tof, tot)]
+        
+    def __calc_trig_chunks_split_indices(self, shot):
+        _, unique_trig_nr_indices, unique_trig_nr_counts = np.unique(shot, return_index=True, return_counts=True)
+        
+        trigger_chunks = []
+        trigger_chunk_voxel_counter = 0
+        for index, unique_trig_nr_index in enumerate(unique_trig_nr_indices):
+            if trigger_chunk_voxel_counter < self._chunk_size_limit:
+                trigger_chunk_voxel_counter += unique_trig_nr_counts[index]
+            else:
+                trigger_chunks.append(unique_trig_nr_index)
+                trigger_chunk_voxel_counter = unique_trig_nr_counts[index]
+
+        return trigger_chunks
+
+    def __centroid_chunks_to_centroids(self, chunks):
+        centroids = [[] for i in range(7)]
```

Comment:
Because in this case this is the number of dimensions which is determined by the `calculate_centroids_properties`-method. This builds up the stack of columns for the centroids (which is initially empty).

This implementation might be bad (not sure) and I can not remember why np.concatenate(chunks) did not work here but there must have been some reason for this. I should check this again i guess.

Reply:
Very nice! Was able to replace the complete function with np.concatenate(list(chunks), axis=1). Should be easier to understand and faster at the same time. I will also remove the function call and use the function directly instead.